### PR TITLE
Added Wiener index

### PR DESCRIPTION
**Pull request by chmurzak**

### CODE REVIEW

File: skfp/fingerprints/weiner_index.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+from scipy.sparse import csr_array
+
+from skfp.bases import BaseFingerprintTransformer
+from skfp.utils import ensure_mols
+
+
+class WeinerIndexFingerprint(BaseFingerprintTransformer):
+    """
+    Weiner Index fingerprint.
+
```
Comment (j-adamczyk) [inline skfp/fingerprints/weiner_index.py:15]: This shouldn't be a fingerprint, this is a single descriptor. It is, however, a part of other descriptors, and a useful thing to have.

Create `descriptors` directory next to `fingerprints`, and maybe `topological.py` file there, and put Wiener index in there. Also, it should be a function `wiener_index(mol, distance_matrix: Optional[np.ndarray] = None) -> int`. If user has a precomputed distance matrix, it should be reused.

Comment (j-adamczyk) [reply skfp/fingerprints/weiner_index.py:15]: Also, add the appropriate citation. See other fingerprints for a format.

File: skfp/fingerprints/weiner_index.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    def _calculate_fingerprint(
+        self, X: Sequence[Union[str, Mol]]
+    ) -> Union[np.ndarray, csr_array]:
+
+        X = ensure_mols(X)
+
+        weiner_indices = []
+        for mol in X:
+            if mol is None:
+                weiner_indices.append(0.0)
```
Comment (j-adamczyk) [inline skfp/fingerprints/weiner_index.py:84]: This should throw an error

File: skfp/fingerprints/weiner_index.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                continue
+
+            distance_matrix = GetDistanceMatrix(mol)
+
+            weiner_index = np.sum(distance_matrix) / 2
```
Comment (j-adamczyk) [inline skfp/fingerprints/weiner_index.py:89]: This can be `// 2`, since sum of distances will always be an integer

File: tests/fingerprints/weiner_index.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import numpy as np
+from rdkit.Chem import GetDistanceMatrix
+from scipy.sparse import csr_array
+
+from skfp.fingerprints import WeinerIndexFingerprint
+
+
```
Comment (j-adamczyk) [inline tests/fingerprints/weiner_index.py:7]: This requires more tests with known values, at least:
1. Examples from the original Wiener's paper
2. Single atom molecule
3. Linear molecule, e.g. butane
4. Ring-like molecule, e.g. benzene

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

**Comment by iedylag** [comment]
Changes after review:
 1. Created descriptors directory and topological.py file 
 2. Added few descriptors as functions (reusing precomputed distance matrix)
 3. Added tests with knows values

**Comment by mjste** [comment]
* I think descriptors should be rather a part of the library, so directory descriptors should become skfp/descriptors
* Some docstring for Topologcal Descriptors missing
* Sync with master

File: .pre-commit-config.yaml
Comment (mjste) [inline .pre-commit-config.yaml:1]: Leave line endings as they were

File: descriptors/topological.py
Comment (mjste) [inline descriptors/topological.py:1]: Move it to skfp/descriptors. Otherwise they will be not available in library

**Review by mjste — CHANGES_REQUESTED** [review]

File: tests/descriptors/topological.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    ],
+)
+def test_descriptors(diverse_molecules, mol_name, descriptor, expected_value):
+    mol_tuple = next(item for item in diverse_molecules if item[0] == mol_name)
+    mol, distance_matrix = mol_tuple[1], mol_tuple[2]
+    topological_descriptor = TopologicalDescriptors(mol, distance_matrix)
+
+    result = getattr(topological_descriptor, descriptor)()
+
+    assert result == pytest.approx(expected_value, abs=1e-6)
```
Comment (j-adamczyk) [inline tests/descriptors/topological.py:63]: Also, add tests that check both provided and not provided `distance_matrix`

**Review by j-adamczyk — CHANGES_REQUESTED** [review]
I also second all comments by @mjste, and added my own

**Comment by iedylag** [comment]
Changes after review:

- Moved descriptors to skfp
- Refactored descriptors to functions
- Added appropriate docstrings with citation
- Added separate tests

File: skfp/descriptors/__init__.py
Comment (j-adamczyk) [inline skfp/descriptors/__init__.py:1]: This file should import descriptors from their module

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -1,71 +1,71 @@
-repos:
```
Comment (j-adamczyk) [inline .pre-commit-config.yaml:1]: Why change line endings? Revert this

File: skfp/descriptors/topological.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    ----------
+    mol : RDKit Mol object
+        The molecular graph for which the Wiener index is to be calculated.
+
+    distance_matrix : np.ndarray, optional
+        Precomputed distance matrix. If not provided, it will be calculated using RDKit.
+
+    References
+    ----------
+    .. [1] Chen, Ailian, Xianzhu Xiong, and Fenggen Lin.
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:26]: 1. Why this reference? The original one is https://pubs.acs.org/doi/abs/10.1021/ja01193a005, second one could be e.g. https://www.sciencedirect.com/science/article/abs/pii/B9781898563761500068.
2. Missing backticks and underscore at the end, see e.g. https://github.com/scikit-fingerprints/scikit-fingerprints/blob/master/skfp/fingerprints/atom_pair.py#L119. This will create a hyperlink in the docs

File: skfp/descriptors/topological.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    Wiener Index (W).
+
+    The implementation uses RDKit to calculate the Weiner Index for molecular graphs.
+    The Weiner Index is computed as the sum of all pairwise distances in the molecular
+    graph's distance matrix.
+
+    Parameters
+    ----------
+    mol : RDKit Mol object
+        The molecular graph for which the Wiener index is to be calculated.
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:19]: Just `The molecule`, since it's quite more than just a graph

File: skfp/descriptors/topological.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    References
+    ----------
+    .. [1] Chen, Ailian, Xianzhu Xiong, and Fenggen Lin.
+        "Explicit relation between the Wiener index and the edge-Wiener index of the catacondensed hexagonal systems."
+        Applied Mathematics and Computation 273 (2016): 1100-1106.
+        <https://www.sciencedirect.com/science/article/pii/S0096300315014149>
+
+    Examples
+    --------
+    >>> from rdkit import Chem
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:33]: Import `MolFromSmiles` directly

File: skfp/descriptors/topological.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        distance_matrix = GetDistanceMatrix(mol)
+    return np.sum(distance_matrix) // 2
+
+
+def average_weiner_index(mol: Mol, distance_matrix: Optional[np.ndarray] = None) -> float:
+    """
+    Average Wiener Index (AW).
+
+    The implementation uses RDKit to calculate the Average Wiener Index for molecular graphs.
+    It is derived by dividing the Wiener index by the total number of atom pairs in the molecule.
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:49]: Maybe also note that this makes it independent of the molecule size.

File: skfp/descriptors/topological.py (hunk 8 above)
Comment (j-adamczyk) [reply skfp/descriptors/topological.py:19]: It applies to all descriptions

File: skfp/descriptors/topological.py (hunk 7 above)
Comment (j-adamczyk) [reply skfp/descriptors/topological.py:26]: Second point applies to all docstrings

File: skfp/descriptors/topological.py (hunk 9 above)
Comment (j-adamczyk) [reply skfp/descriptors/topological.py:33]: Applies to all docstrings

File: skfp/descriptors/topological.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+        <https://www.sciencedirect.com/science/article/pii/S0097848599000352>
+
+    Examples
+    --------
+    >>> from rdkit import Chem
+    >>> from descriptors.topological import average_weiner_index
+    >>> mol = Chem.MolFromSmiles("C1=CC=CC=C1")  # Benzene
+    >>> average_weiner_index(mol)
+    1.8
+"""
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:73]: Missing tab

File: skfp/descriptors/topological.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    mol : RDKit Mol object
+        The molecular graph for which the Graph Distance Index is to be calculated.
+
+    distance_matrix : np.ndarray, optional
+        Precomputed distance matrix. If not provided, it will be calculated using RDKit.
+
+    References
+    ----------
+    .. [1] PyChem documentation
+        <https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/pychem/manual.pdf>
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:110]: Proper reference is `Rouvray, D. H. In Chemical Applications of Topology and Graph Theory; King, R. B., Ed.; Elsevier: Amsterdam, 1983; pp 159-177.`, or https://pubs.acs.org/doi/10.1021/ci9502461

File: skfp/descriptors/topological.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+    Examples
+    --------
+    >>> from rdkit import Chem
+    >>> from descriptors.topological import graph_distance_index
+    >>> mol = Chem.MolFromSmiles("C1=CC=CC=C1")  # Benzene
+    >>> graph_distance_index(mol)
+    261.0
+"""
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:119]: Missing tab

File: skfp/descriptors/topological.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+        The molecular graph for which the Zagreb index is to be calculated.
+
+    adjacency_matrix : np.ndarray, optional
+        Precomputed adjacency matrix. If not provided, it will be calculated using RDKit.
+
+    References
+    ----------
+    .. [1] Gutman, Ivan.
+        "Degree-based topological indices."
+        Croatica chemica acta 86.4 (2013): 352.
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:146]: Should be capitalized, i.e. Croatica Chemica Acta

File: skfp/descriptors/topological.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+    >>> from descriptors.topological import zagreb_index
+    >>> mol = Chem.MolFromSmiles("C1=CC=CC=C1")  # Benzene
+    >>> zagreb_index(mol)
+    24
+    """
+    if adjacency_matrix is None:
+        adjacency_matrix = rdmolops.GetAdjacencyMatrix(mol)
+
+    degree_sequence = adjacency_matrix.sum(axis=1)
+    return sum(degree_sequence ** 2)
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:161]: It's cheaper to do:
```
sum(atom.GetDegree() ** 2 for atom in mol.GetAtoms())
```

File: skfp/descriptors/topological.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+    carbon_only : bool, default=False
+        Whether to consider only carbon-carbon distances. If `True`, the distance matrix will be filtered
+        to include only rows and columns corresponding to carbon atoms.
+
+    References
+    ----------
+    .. [1] Imran, Muhammad, Mehar Ali Malik, and Ramsha Javed.
+        "Wiener polarity index and related molecular topological descriptors of titanium oxide nanotubes."
+        International Journal of Quantum Chemistry 121.11 (2021): e26627.
+        <https://onlinelibrary.wiley.com/doi/abs/10.1002/qua.26627>
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:191]: This is a good second reference, but use original Wiener index paper as first reference

File: skfp/descriptors/topological.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+    >>> mol = Chem.MolFromSmiles("C1=CC=CC=C1")  # Benzene
+    >>> polarity_number(mol)
+    3
+"""
+    if distance_matrix is None:
+        distance_matrix = GetDistanceMatrix(mol)
+
+    if carbon_only:
+        atom_indices = [
+            i for i, atom in enumerate(mol.GetAtoms()) if atom.GetAtomicNum() == 6  # Carbon atomic number is 6
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:206]: You can also do `.GetSymbol() == "C"`

File: tests/descriptors/topological.py [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+    "mol_name, expected_value",
+    [
+        ("ethane", 1.0),
+        ("ethanol", 1.3333333333333333),
+        ("carbon_dioxide", 1.3333333333333333),
+        ("benzene", 1.8),
+        ("acetic_acid", 1.5),
+        ("pyridine", 1.8),
+        ("isobutane", 1.5),
+        ("pyrimidine", 1.9642857142857142),
```
Comment (j-adamczyk) [inline tests/descriptors/topological.py:56]: Round this to e.g. 3rd decimal place and then compare, it's more readable and still accurate enough

File: tests/descriptors/topological.py [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+        ("benzene", 27),
+        ("acetic_acid", 9),
+        ("pyridine", 27),
+        ("isobutane", 9),
+        ("pyrimidine", 55),
+    ],
+)
+def test_weiner_index(mol_name, expected_value):
+    mol_tuple = next(item for item in diverse_molecules() if item[0] == mol_name)
+    mol = mol_tuple[1]
```
Comment (j-adamczyk) [inline tests/descriptors/topological.py:35]: Just change `diverse_molecules` to return `dict[str, Mol]` and then you can simply select by name

Comment (j-adamczyk) [reply tests/descriptors/topological.py:35]: In fact, this should simply all tests here

File: tests/descriptors/topological.py [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+    return [
+        ("ethane", MolFromSmiles("CC")),
+        ("ethanol", MolFromSmiles("CCO")),
+        ("carbon_dioxide", MolFromSmiles("O=C=O")),
+        ("benzene", MolFromSmiles("C1=CC=CC=C1")),
+        ("acetic_acid", MolFromSmiles("CC(=O)O")),
+        ("pyridine", MolFromSmiles("c1ccncc1")),
+        ("isobutane", MolFromSmiles("C(C)(C)C")),
+        ("pyrimidine", MolFromSmiles("c1cnc2ncnc12")),
+    ]
```
Comment (j-adamczyk) [inline tests/descriptors/topological.py:17]: You can also calculate the distance matrices here, and use them in tests: first without passing it (descriptor function computing it internally), then passing. Both results should be the same.

File: skfp/descriptors/topological.py [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+"""
+
+    if distance_matrix is None:
+        distance_matrix = GetDistanceMatrix(mol)
+    wiener_idx = weiner_index(mol, distance_matrix)
+    num_atoms = mol.GetNumAtoms()
+    return (2 * wiener_idx) / (num_atoms * (num_atoms - 1))
+
+
+def graph_distance_index(mol: Mol, distance_matrix: Optional[np.ndarray] = None) -> float:
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:82]: This returns `int`

Comment (j-adamczyk) [reply skfp/descriptors/topological.py:82]: Or at least it should, cast if necessary

File: skfp/descriptors/topological.py [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+
+    Parameters
+    ----------
+    mol : RDKit Mol object
+        The molecular graph for which the Polarity Number is to be calculated.
+
+    distance_matrix : np.ndarray, optional
+        Precomputed distance matrix. If not provided, it will be calculated using RDKit.
+
+    carbon_only : bool, default=False
```
Comment (j-adamczyk) [inline skfp/descriptors/topological.py:182]: Edge case - what if molecules does not have any carbons? This should raise an explicit error then. You can check this, i.e. if `atom_indices` is empty or not. Also, write a test for this.

File: tests/descriptors/topological.py [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+)
+def test_polarity_number_carbon_only(mol_name, expected_value):
+    mol_tuple = next(item for item in diverse_molecules() if item[0] == mol_name)
+    mol = mol_tuple[1]
+
+    distance_matrix = GetDistanceMatrix(mol)
+
+    result = top.polarity_number(mol, distance_matrix, carbon_only=True)
+
+    assert (result == pytest.approx(expected_value, abs=1e-6))
```
Comment (j-adamczyk) [inline tests/descriptors/topological.py:164]: Why additional parentheses?

File: tests/descriptors/topological.py [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+def test_weiner_index(mol_name, expected_value):
+    mol_tuple = next(item for item in diverse_molecules() if item[0] == mol_name)
+    mol = mol_tuple[1]
+
+    distance_matrix = GetDistanceMatrix(mol)
+
+    result = top.weiner_index(mol, distance_matrix)
+    result_without_distance_matrix = top.weiner_index(mol)
+
+    assert result == pytest.approx(expected_value, abs=1e-6)
```
Comment (j-adamczyk) [inline tests/descriptors/topological.py:42]: You are comparing integers here, no need for approximate testing, since it's much slower

**Review by j-adamczyk — CHANGES_REQUESTED** [review]
Also add docs in `docs` directory, with appropriate `.rst` files

File: .pre-commit-config.yaml (hunk 6 above)
Comment (iedylag) [reply .pre-commit-config.yaml:1]: I changed this in previous commit. Couldn't revert because of the merge with master, so i copied the content of this file from the master branch. This is the current version of the file.

File: skfp/descriptors/topological.py (hunk 16 above)
Comment (iedylag) [reply skfp/descriptors/topological.py:191]: Are any of these suitable for first reference?
1. https://pubs.acs.org/doi/10.1021/ci970122j
2. https://match.pmf.kg.ac.rs/electronic_versions/Match66/n1/match66n1_293-304.pdf

File: tests/descriptors/topological.py (hunk 24 above)
Comment (iedylag) [reply tests/descriptors/topological.py:42]: I didn't know it would slow down the tests. I removed it from every test because it wasn't necessary in any test case.

**Review by j-adamczyk — APPROVED** [review]
I made a few smaller changes myself. Also note the changes in the documentation.

**Review by j-adamczyk — APPROVED** [review]

**Review by j-adamczyk — APPROVED** [review]

**Review by mjste — APPROVED** [review]
